File: leetcode/442/442.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int leng = nums.size();
        vector<int> dupes;
        int last = 1;
        for(int i = 0; i < leng; i++){
            // length test
            if(abs(nums[i]) == leng){
                if(last < 0){
                    dupes.push_back(leng);
                } else {
                    last = -last;
                }
            } else if(nums[abs(nums[i])] < 0){
                dupes.push_back(abs(nums[i]));
            } else {
                nums[abs(nums[i])] = -nums[abs(nums[i])];
            }
        }
        return dupes;
    }
};

File: leetcode/problems/442/442.py
class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ht = dict();
        duplicates = []
        for num in nums:
            if num in ht:
                duplicates.append(num)
            else:
                ht[num] = 1
        return duplicates
        